internal/db/db.go

```diff
@@ -121,6 +149,37 @@ func (db *DB) waitUpgrade(bootstrap bool) error {
 
 	err := db.retry(context.TODO(), func(_ context.Context) error {
 		_, err := newSchema.Ensure(db.db)
+		if err != nil {
+			return err
+		}
+
+		if !bootstrap {
+			// Perform the API extensions check.
+			err = query.Transaction(context.TODO(), db.db, func(ctx context.Context, tx *sql.Tx) error {
+				err := cluster.UpdateClusterMemberAPIExtensions(tx, ext, db.listenAddr.URL.Host)
+				if err != nil {
+					return fmt.Errorf("Failed to update API extensions when joining cluster: %w", err)
+				}
+
+				clusterMembersAPIExtensions, err := cluster.GetClusterMemberAPIExtensions(ctx, tx)
+				if err != nil {
+					return fmt.Errorf("Failed to get other members' API extensions: %w", err)
+				}
+
+				otherNodesBehindInternal, err := checkAPIExtensions(ext, clusterMembersAPIExtensions)
+				if err != nil {
+					return err
+				}
+
+				if otherNodesBehindInternal {
+					otherNodesBehind = true
+					return schema.ErrGracefulAbort
+				}
+
+				return nil
+			})
+		}
+
 		return err
```

Comment:
This was a sneaky error! If you return `schema.ErrGracefulAbort` from within the transaction, then the transaction will never be committed and when each node updates its local API version in the database, it won't be reflected on other nodes.

You can fix it by moving the check for `otherNodesBehind` outside the transaction.

```suggestion
		if !bootstrap {
		        otherNodesBehindAPI := false
			// Perform the API extensions check.
			err = query.Transaction(context.TODO(), db.db, func(ctx context.Context, tx *sql.Tx) error {
				err := cluster.UpdateClusterMemberAPIExtensions(tx, ext, db.listenAddr.URL.Host)
				if err != nil {
					return fmt.Errorf("Failed to update API extensions when joining cluster: %w", err)
				}

				clusterMembersAPIExtensions, err := cluster.GetClusterMemberAPIExtensions(ctx, tx)
				if err != nil {
					return fmt.Errorf("Failed to get other members' API extensions: %w", err)
				}

				otherNodesBehindAPI, err := checkAPIExtensions(ext, clusterMembersAPIExtensions)
				if err != nil {
					return err
				}

				return nil
			})
			if err != nil {
			  return err
			}
			
			if otherNodesBehindInternal {
				otherNodesBehind = true
				return schema.ErrGracefulAbort
		        }
		}

		return nil
```

Reply:
Also, I think we should halve the timeout here: https://github.com/canonical/microcluster/blob/94483f4d619243d598d4b5e512735035d1e34445/internal/db/db.go#L132

The reason for that is because if an API and schema update happen simultaneously, we'll have to wait up to 2 whole minutes for each node to pick up the change.

Reply:
You also may run into the bug fixed by #117 so you may want to rebase off of that to test locally.

Reply:
Amazing! thanks a lot!

Reply:
fixed